### PR TITLE
[SPARK-45056][PYTHON][SS][CONNECT] Termination tests for streamingQueryListener and foreachBatch

### PR DESCRIPTION
**Pull request by WweiL**
### What changes were proposed in this pull request?

Add termination tests for StreamingQueryListener and foreachBatch. 

The behavior is mimicked by creating the same query on server side that would have been created if running the same python query is ran on client side. For example, in foreachBatch, a python foreachBatch function is serialized using cloudPickleSerializer and passed to the server side, here we start another python process on the server and call the same cloudPickleSerializer and pass the bytes to the server, and construct `SimplePythonFunction` accordingly.

Refactored the code a bit for testing purpose.

### Why are the changes needed?

Necessary tests

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Test only addition

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by WweiL** [comment]
CC @bogao007 @rangadi @ueshin Can you guys take a look? Thanks in advance!

**Comment by WweiL** [comment]
Reminder to myself: https://github.com/apache/spark/pull/42687 and this PR will have conflicts in `streamingForeachBatchRunnerCleanerCache`

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala [hunk 1]
```diff
@@ -79,4 +90,135 @@ class SparkConnectSessionHolderSuite extends SharedSparkSession {
...
+      envVars = mutable.Map("PYTHONPATH" -> sparkPythonPath).asJava,
+      pythonIncludes = sessionHolder.artifactManager.getSparkConnectPythonIncludes.asJava,
+      pythonExec = IntegratedUDFTestUtils.pythonExec,
+      pythonVer = IntegratedUDFTestUtils.pythonVer,
+      broadcastVars = Lists.newArrayList(),
+      accumulator = null)
+  }
+
+  test("python foreachBatch process: process terminates after query is stopped") {
+    val sessionHolder = SessionHolder.forTesting(spark)
```
Comment (rangadi) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:111]: How does this test if the actual python process is not running?

File: python/pyspark/sql/tests/connect/streaming/worker_for_testing.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+A worker for streaming foreachBatch in Spark Connect.
+Usually this is ran on the driver side of the Spark Connect Server.
```
Comment (rangadi) [inline python/pyspark/sql/tests/connect/streaming/worker_for_testing.py:20]: Why do we need this? better to add description of that here.

File: python/pyspark/sql/tests/connect/streaming/worker_for_testing.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    session_id = utf8_deserializer.loads(infile)
+
+    print(
+        "Streaming test worker is starting with " f"url {connect_url} and sessionId {session_id}."
+    )
+
+    write_int(0, outfile)  # Indicate successful initialization
+    outfile.flush()
+
+    while True:
```
Comment (rangadi) [inline python/pyspark/sql/tests/connect/streaming/worker_for_testing.py:50]: Busy loop like can add a lot of burden on the test infra and make tests flaky

Comment (WweiL) [reply python/pyspark/sql/tests/connect/streaming/worker_for_testing.py:50]: Thanks. Removed the test worker. Found a way to test the actual python code.

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala (hunk 1 above)
Comment (WweiL) [reply connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:111]: Found out a new way of testing this. Please see the update

**Review by HyukjinKwon — APPROVED** [review]
LGTM, pending tests.

**Comment by WweiL** [comment]
Now it fails with `No module named 'grpc_status'`, let me search through how to resolve this

File: .github/workflows/build_and_test.yml [hunk 4]
```diff
@@ -256,14 +256,14 @@ jobs:
...
+      if: contains(matrix.modules, 'yarn') || (contains(matrix.modules, 'sql') && !contains(matrix.modules, 'sql-')) || contains(matrix.modules, 'connect')
       with:
         python-version: 3.8
         architecture: x64
     - name: Install Python packages (Python 3.8)
-      if: (contains(matrix.modules, 'sql') && !contains(matrix.modules, 'sql-'))
+      if: (contains(matrix.modules, 'sql') && !contains(matrix.modules, 'sql-')) || contains(matrix.modules, 'connect')
       run: |
-        python3.8 -m pip install 'numpy>=1.20.0' pyarrow pandas scipy unittest-xml-reporting 'grpcio==1.56.0' 'protobuf==3.20.3'
+        python3.8 -m pip install 'numpy>=1.20.0' pyarrow pandas scipy unittest-xml-reporting 'grpcio==1.56.0' 'protobuf==3.20.3' grpcio-status
```
Comment (HyukjinKwon) [inline .github/workflows/build_and_test.yml:266]:
```suggestion
        python3.8 -m pip install 'numpy>=1.20.0' pyarrow pandas scipy unittest-xml-reporting 'grpcio==1.56.0' 'protobuf==3.20.3' grpcio-status=='1.56.0'
```

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala [hunk 5]
```diff
@@ -79,4 +92,196 @@ class SparkConnectSessionHolderSuite extends SharedSparkSession {
...
+        // assert query2's python process is not running
+        assert(runner2.isWorkerStopped().get)
+      }
+
+      assert(spark.streams.active.isEmpty) // no running query
+      assert(spark.streams.listListeners().length == 1) // only process termination listener
+    } finally {
+      SparkConnectService.stop()
+      // remove process termination listener
+      spark.streams.removeListener(spark.streams.listListeners()(0))
```
Comment (yaooqinn) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:234]: Hi @WweiL @HyukjinKwon,  we encountered a test failure with this line 

```scala
python foreachBatch process: process terminates after query is stopped *** FAILED *** (1 second, 431 milliseconds)
[info]   java.lang.ArrayIndexOutOfBoundsException: 0
[info]   at org.apache.spark.sql.connect.service.SparkConnectSessionHolderSuite.$anonfun$new$7(SparkConnectSessionHodlerSuite.scala:234)
```

How about changing it to
```scala
spark.streams.listListeners().foreach(spark.streams.removeListener)
```

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala [hunk 6]
```diff
@@ -79,4 +92,196 @@ class SparkConnectSessionHolderSuite extends SharedSparkSession {
...
+
+    val sessionHolder = SessionHolder.forTesting(spark)
+    try {
+      SparkConnectService.start(spark.sparkContext)
+
+      val pythonFn = dummyPythonFunction(sessionHolder)(streamingQueryListenerFunction)
+
+      val id1 = "listener_removeListener_test_1"
+      val id2 = "listener_removeListener_test_2"
+      val listener1 = new PythonStreamingQueryListener(pythonFn, sessionHolder)
```
Comment (yaooqinn) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:251]: this test also crushed at this line
```
[info] - python listener process: process terminates after listener is removed *** FAILED *** (427 milliseconds)
[info]   java.io.EOFException:
[info]   at java.io.DataInputStream.readInt(DataInputStream.java:392)
[info]   at org.apache.spark.api.python.StreamingPythonRunner.init(StreamingPythonRunner.scala:101)
[info]   at org.apache.spark.sql.connect.planner.PythonStreamingQueryListener.<init>(StreamingQueryListenerHelper.scala:41)
[info]   at org.apache.spark.sql.connect.service.SparkConnectSessionHolderSuite.$anonfun$new$11(SparkConnectSessionHodlerSuite.scala:251)
[info]   at org.scalatest.enablers.Timed$$anon$1.timeoutAfter(Timed.scala:127)
[info]   at org.scalatest.concurrent.TimeLimits$.failAfterImpl(TimeLimits.scala:282)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter(TimeLimits.scala:231)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter$(TimeLimits.scala:230)
```

**Comment by HyukjinKwon** [comment]
let me make a quick followup or revert.

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala (hunk 6 above)
Comment (HyukjinKwon) [reply connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:251]: I think this was a flakiness.

File: connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingQueryListenerHelper.scala [hunk 7]
```diff
@@ -26,15 +26,13 @@ import org.apache.spark.sql.streaming.StreamingQueryListener
...
-    listener: SimplePythonFunction,
-    sessionHolder: SessionHolder,
-    pythonExec: String)
+class PythonStreamingQueryListener(listener: SimplePythonFunction, sessionHolder: SessionHolder)
     extends StreamingQueryListener {
 
   private val port = SparkConnectService.localPort
   private val connectUrl = s"sc://localhost:$port/;user_id=${sessionHolder.userId}"
-  private val runner = StreamingPythonRunner(
+  // Scoped for testing
```
Comment (rangadi) [inline connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/StreamingQueryListenerHelper.scala:34]: Nit: _// Visible for testing._

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala [hunk 8]
```diff
@@ -79,4 +92,196 @@ class SparkConnectSessionHolderSuite extends SharedSparkSession {
       sessionHolder.getDataFrameOrThrow(key1)
     }
   }
+
+  private def streamingForeachBatchFunction(pysparkPythonPath: String): Array[Byte] = {
```
Comment (rangadi) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:96]: Add a brief comment about what this does (and/or it purpose).

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala [hunk 9]
```diff
@@ -79,4 +92,196 @@ class SparkConnectSessionHolderSuite extends SharedSparkSession {
...
+      val query2 = spark.readStream
+        .format("rate")
+        .load()
+        .writeStream
+        .format("memory")
+        .queryName("foreachBatch_termination_test_q2")
+        .foreachBatch(fn2)
+        .start()
+
+      sessionHolder.streamingForeachBatchRunnerCleanerCache
```
Comment (rangadi) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:204]: Add comment about why we are accessing internal cleaner cache here.
It wouldn't be clear to me if you hadn't explain in VC. Refer to how this this mimicking Python foreachBatch API from Scala.

File: core/src/main/scala/org/apache/spark/api/python/StreamingPythonRunner.scala [hunk 10]
```diff
@@ -108,10 +108,30 @@ private[spark] class StreamingPythonRunner(
...
+      }
+    } catch {
+      case e: Exception =>
+        logError("Exception when trying to kill worker", e)
+    }
+  }
+
+  /**
+   * Returns whether the Python worker has been stopped.
+   * @return Some(true) if the Python worker has been stopped.
```
Comment (rangadi) [inline core/src/main/scala/org/apache/spark/api/python/StreamingPythonRunner.scala:128]: Simpler return `Boolean` and throw if there is no worker started. This is only used in tests.

**Review by rangadi — COMMENTED** [review]
Made a few suggestions, then realized this is already closed :). 
No big deal.

**Comment by HyukjinKwon** [comment]
@WweiL mind creating a followup to fix them?

File: connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala [hunk 11]
```diff
@@ -79,4 +92,196 @@ class SparkConnectSessionHolderSuite extends SharedSparkSession {
...
+      assert(spark.streams.active.isEmpty) // no running query
+      assert(spark.streams.listListeners().length == 1) // only process termination listener
+    } finally {
+      SparkConnectService.stop()
+      // remove process termination listener
+      spark.streams.removeListener(spark.streams.listListeners()(0))
+    }
+  }
+
+  test("python listener process: process terminates after listener is removed") {
```
Comment (LuciferYang) [inline connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:238]: @WweiL The newly added test case caused Maven's daily testing to fail. Could you help fix it? Thanks

Comment (LuciferYang) [reply connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:238]: It seems that the new tesgt is needed is `pandas`, so should we change to check

```
assume(IntegratedUDFTestUtils.shouldTestPandasUDFs)
```
?

Comment (LuciferYang) [reply connector/connect/server/src/test/scala/org/apache/spark/sql/connect/service/SparkConnectSessionHodlerSuite.scala:238]: change the assume condition in https://github.com/apache/spark/pull/42976. Afterwards, we should install Python test dependencies for Maven's daily test

**Pull request closed without merging**